controllers/database_reconciler_helpers.go

```diff
@@ -242,62 +201,76 @@ func (r *DatabaseReconciler) handleSync(ctx context.Context, database *ndbv1alph
+		databaseStatus.IPAddress = dbInfo.IPAddress
+		databaseStatus.DatabaseServerId = dbInfo.DBServerId
+	} else if databaseStatus.Status == common.DATABASE_CR_STATUS_WAITING {
+		log.Info("Database not found in NDB CR yet")
+		r.recorder.Event(database, "Normal", EVENT_WAITING_FOR_NDB_RECONCILE, "Waiting for NDB server resource to reconcile")
+	} else {
+		log.Info("Database missing from NDB CR")
+		databaseStatus.Status = common.DATABASE_CR_STATUS_NOT_FOUND
+		// TODO: Add operation tracking while the DB is in PROVISIONING state
+	}
+
+	if !reflect.DeepEqual(database.Status, *databaseStatus) {
+		database.Status = *databaseStatus
+		err := r.Status().Update(ctx, database)
 		if err != nil {
 			errStatement := "Failed to update status of database custom resource"
 			log.Error(err, errStatement)
 			r.recorder.Eventf(database, "Warning", EVENT_CR_STATUS_UPDATE_FAILED, "Error: %s. %s.", err.Error())
 			return requeueOnErr(err)
 		}
+	}
 
-	case common.DATABASE_CR_STATUS_PROVISIONING:
-		// Check the status of the DB
-		databaseResponse, err := ndb_api.GetDatabaseById(ctx, ndbClient, database.Status.Id)
-		if err != nil {
-			errStatement := fmt.Sprintf("Failed to get db with id %s", database.Status.Id)
-			log.Error(err, errStatement)
-			r.recorder.Eventf(database, "Warning", EVENT_RESOURCE_LOOKUP_ERROR, "Error: %s, %s", errStatement, err.Error())
-			return requeueOnErr(err)
-		}
-
-		// if READY => Change status
-		// log.Info("DEBUG Database Response: " + util.ToString(databaseResponse))
-		if databaseResponse.Status == common.DATABASE_CR_STATUS_READY {
-			log.Info("Database instance is READY, adding data to CR's status and updating the CR")
-			r.recorder.Event(database, "Normal", EVENT_PROVISIONING_COMPLETED, "Database has been provisioned on NDB.")
-			database.Status.Status = common.DATABASE_CR_STATUS_READY
-			database.Status.DatabaseServerId = databaseResponse.DatabaseNodes[0].DatabaseServerId
-			database.Status.IPAddress = databaseResponse.DatabaseNodes[0].DbServer.IPAddresses[0]
-			if database.Status.IPAddress != "" {
-				err = r.Status().Update(ctx, database)
-				if err != nil {
-					errStatement := "Failed to update status of database custom resource"
-					log.Error(err, errStatement)
-					r.recorder.Eventf(database, "Warning", EVENT_CR_STATUS_UPDATE_FAILED, "Error: %s. %s.", err.Error())
-					return requeueOnErr(err)
-				}
+	// Handle Internal Sync -
+	// [READY]
+	// Add finalizers only when the database is in ready state so that if
+	// any failure occurrs before reaching the ready state, the failure
+	// would not cause the deletion to block the terminal.
+	// Also, setup and create network services.
+	// [DELETING]
+	// Delete the database instance and the VM as per the finalizers
+	// [NOT FOUND]
+	// Record an event and then do not requeue since the resource has been deleted externally
+	// or was not found on NDB
+	switch databaseStatus.Status {
+	case common.DATABASE_CR_STATUS_READY:
```

Comment:
Thanks for tackling the finalizers

Reply:
what are finalizers?

Reply:
@yash-ntnx 
Finalizers is a list of strings that is attached to the Custom Resource to handle the deletion of external resources owned by the CR. We leverage these finalizers to perform deletion of the Database Instance and the Database Server VM.

From the [Operator SDK docs](https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#external-resources)- 
Sometimes external resources or resources that are not owned by a custom resource, those across namespaces for example, need to be cleaned up when the parent resource is deleted. In that case [Finalizers](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers) can be leveraged. A deletion request for an object with a finalizer becomes an update during which a deletion timestamp is set; the object is not deleted while the finalizer is present. The reconciliation loop of the custom resource’s controller will then need to check whether a the deletion timestamp is set, perform the external cleanup operation(s), then remove the finalizer to allow garbage collection of the object. Multiple finalizers may be present on an object, each with a key that should indicate what external resources require deletion by the controller.